include/sys/dsl_crypt.h

```diff
@@ -0,0 +1,181 @@
+/*
+ * CDDL HEADER START
+ *
+ * The contents of this file are subject to the terms of the
+ * Common Development and Distribution License, Version 1.0 only
+ * (the "License").  You may not use this file except in compliance
+ * with the License.
+ *
+ * You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
+ * or http://www.opensolaris.org/os/licensing.
+ * See the License for the specific language governing permissions
+ * and limitations under the License.
+ *
+ * When distributing Covered Code, include this CDDL HEADER in each
+ * file and include the License file at usr/src/OPENSOLARIS.LICENSE.
+ * If applicable, add the following below this CDDL HEADER, with the
+ * fields enclosed by brackets "[]" replaced with your own identifying
+ * information: Portions Copyright [yyyy] [name of copyright owner]
+ *
+ * CDDL HEADER END
+ */
+/*
+ * Copyright (c) 2016, Datto, Inc. All rights reserved.
+ */
+
+#ifndef	_SYS_DSL_CRYPT_H
+#define	_SYS_DSL_CRYPT_H
+
+#include <sys/dmu_tx.h>
+#include <sys/dmu.h>
+#include <sys/zio_crypt.h>
+#include <sys/spa.h>
+#include <sys/dsl_dataset.h>
+
+/* forward declarations */
+struct dsl_dataset;
+
+typedef enum zfs_keystatus {
+	ZFS_KEYSTATUS_NONE = 0,
+	ZFS_KEYSTATUS_UNAVAILABLE,
+	ZFS_KEYSTATUS_AVAILABLE,
+} zfs_keystatus_t;
+
+/* in memory representation of a wrapping key */
+typedef struct dsl_wrapping_key {
+	/* link into the keystore's tree of wrapping keys */
+	avl_node_t wk_avl_link;
+
+	/* actual wrapping key */
+	crypto_key_t wk_key;
+
+	/* refcount of number of dsl_crypto_key_t's holding this struct */
+	refcount_t wk_refcnt;
+
+	/* dsl directory object that owns this wrapping key */
+	uint64_t wk_ddobj;
+} dsl_wrapping_key_t;
+
+/* structure for passing around encryption params from userspace */
+typedef struct dsl_crypto_params {
+	/* command to be executed */
+	zfs_ioc_crypto_cmd_t cp_cmd;
+
+	/* the encryption algorithm */
+	uint64_t cp_crypt;
+
+	/* the salt, if the keysource is of type passphrase */
+	uint64_t cp_salt;
+
+	/* the pbkdf2 iterations, if the keysource is of type passphrase */
+	uint64_t cp_iters;
+
+	/* keysource property string */
+	const char *cp_keysource;
+
+	/* the wrapping key */
+	dsl_wrapping_key_t *cp_wkey;
```

Comment:
Sounds good, I was thinking it was passed from userland like the `zfs_cmd_t`, in which case it would be confusing as to which address space the pointer is in (and it passing from a 32-bit userland to 64-bit kernel wouldn't work right due to differing pointer sizes).